src/crafting.cpp

```diff
@@ -1335,112 +1054,63 @@ void player::consume_tools( const std::vector<tool_comp> &tools, int batch,
-                              const bool print_msg ) const
+bool player::can_disassemble( const item &obj, const inventory &inv, bool alert ) const
 {
-    if( dis_item.is_book() ) {
-        return true;
-    }
-
-    for( auto cur_recipe : recipe_dict ) {
-        if( dis_item.typeId() == cur_recipe->result && cur_recipe->reversible ) {
-            return can_disassemble( dis_item, cur_recipe, crafting_inv, print_msg );
+    const auto &r = recipe_dictionary::get_uncraft( obj.typeId() );
+    if( !r ) {
+        if( alert ) {
+            add_msg_if_player( m_info, _( "You cannot disassemble the %s" ), obj.tname().c_str() );
         }
+        return false;
     }
 
-    return false;
-}
-
-bool player::can_disassemble( const item &dis_item, const recipe *cur_recipe,
-                              const inventory &crafting_inv, bool print_msg ) const
-{
-    const std::string dis_name = dis_item.tname().c_str();
-    if( dis_item.count_by_charges() && !cur_recipe->has_flag( "UNCRAFT_SINGLE_CHARGE" ) ) {
+    if( obj.count_by_charges() && !r.has_flag( "UNCRAFT_SINGLE_CHARGE" ) ) {
         // Create a new item to get the default charges
-        const item tmp = cur_recipe->create_result();
-        if( dis_item.charges < tmp.charges ) {
-            if( print_msg ) {
-                popup( ngettext( "You need at least %d charge of %s to disassemble it.",
-                                 "You need at least %d charges of %s to disassemble it.",
-                                 tmp.charges ),
-                       tmp.charges, dis_name.c_str() );
+        int qty = r.create_result().charges;
+        if( obj.charges < qty ) {
+            if( alert ) {
+                auto msg = ngettext( "You need at least %d charge of %s to disassemble it.",
+                                     "You need at least %d charges of %s to disassemble it.", qty );
+                add_msg_if_player( m_info, msg, qty, obj.tname().c_str() );
             }
             return false;
         }
     }
 
-    bool have_all_qualities = true;
-    const auto &dis_requirements = cur_recipe->requirements().disassembly_requirements();
-    for( const auto &itq : dis_requirements.get_qualities() ) {
-        for( const auto &it : itq ) {
-            if( !it.has( crafting_inv ) ) {
-                if( print_msg ) {
-                    add_msg( m_info, _( "To disassemble %s, you need %s" ),
-                             dis_name.c_str(), it.to_string().c_str() );
-                }
+    const auto &dis = r.disassembly_requirements();
 
-                have_all_qualities = false;
+    for( const auto &opts : dis.get_qualities() ) {
+        for( const auto &qual : opts ) {
+            if( !qual.has( inv ) ) {
+                if( alert ) {
+                    add_msg_if_player( m_info, _( "To disassemble %s, you need %s" ),
+                                       obj.tname().c_str(), qual.to_string().c_str() );
+                }
+                return false;
```

Comment:
Want to redo qualities to actual support alternatives so will address this in another PR